File: ui/src/study/Study.js

export default class Study {
	constructor(data) {
		this._data = data /* = {
			name: "",
			filepath: "",
			media: [
				{
					id: "",
					name: "",
					tagIds: [""],
					filepath: "",
					sections: [
						{
							id: "",
							name: "",
							tagIds: [""],
							startTime: 0, // in seconds
							endTime: 0, // in seconds
						}
					]
				}
			],
			sessions: [
				{
					id: "",
					name: "",
					tagIds: [""],
					observation: [
						{
							id: "",
							name: "",
							tagIds: [""],
							startTime: 0, // in seconds
							endTime: -1, // in seconds
							quote: "",
							description: "",
						}
					]
				}
			],
			codes: [
				{
					id: "",
					name: "",
					tagIds: [""],
					description: "",
				}
			],
			tags: [
				{
					id: "",
					name: "",
					description: "",
				}
			]
		} */
	}

	get data() {
		return structuredClone(this._data)
	}

	get name() {
		return this._data.name
	}

	get filepath() {
		return this._data.filepath
	}

	get media() {
		return this.toObject().media
	}

	get sessions() {
		return this.toObject().sessions
	}

	get codes() {
		return this.toObject().codes
	}

	get tags() {
		return this.toObject().tags
	}

	// Creates a new tag.
	newTag(name, description="") {
		// TODO
	}

	// Updates an existing tag.
	updateTag(id, { name=null, description=null }) {
		// TODO
	}

	// Deletes an existing tag.
	deleteTag(id) {
		// TODO
	}

	// Searches all data to find the thing (entity, object, entry) with the
	// specified ID, else returns null.
	findThingById(id) {
		// TODO

		return null
	}

	findThingsWithTag(tag) {
		// TODO
	}

	findThingsWithTagId(tagId) {
		// TODO
	}
}

File: api/study/study.go
package study

import (
	"encoding/json"
	"os"
)

func CreateAndSaveNewStudy(name string, path string) (*Study, error) {
	study := Study{
		Name:     name,
		Filepath: path,
		Media:    []Media{},
		Codes:    []Code{},
		Tags:     []Tag{},
	}

	data, e := json.Marshal(study)
	if e != nil {
		return nil, e
	}

	e = os.WriteFile(name+".json", data, 0666)
	if e != nil {
		return nil, e
	}

	return &study, nil
}

type Study struct {
	Name     string  `json:"name"`
	Filepath string  `json:"filepath"`
	Media    []Media `json:"media"`
	Codes    []Code  `json:"code"`
	Tags     []Tag   `json:"tag"`
}

type StudyEntity struct {
	Id   string `json:"id"`
	Name string `json:"name"`
}

type TaggableStudyEntity struct {
	StudyEntity
	Tags string `json:"tags"`
}

type Media struct {
	TaggableStudyEntity
	Filepath     string         `json:"filepath"`
	Sections     []MediaSection `json:"sections"`
	Observations []Observation  `json:"observations"`
}

type MediaSection struct {
	TaggableStudyEntity
	Start int `json:"start"`
	End   int `json:"end"`
}

type Observation struct {
	TaggableStudyEntity
	Start       int      `json:"start"`
	Quote       string   `json:"quote"`
	Description string   `json:"description"`
	Codes       []string `json:"codes"`
}

type Code struct {
	TaggableStudyEntity
	Description string `json:"description"`
}

type Tag struct {
	StudyEntity
	Description string `json:"description"`
}
